include/gridtools/common/hugepage_alloc.hpp

```diff
@@ -11,25 +11,178 @@
+
+        inline hugepage_mode hugepage_mode_from_env() {
+            const char *env_value = std::getenv("GT_HUGEPAGE_MODE");
+            if (!env_value || std::strcmp(env_value, "transparent") == 0)
+                return hugepage_mode::transparent;
+            if (std::strcmp(env_value, "disable") == 0)
+                return hugepage_mode::disabled;
+            if (std::strcmp(env_value, "explicit") == 0)
+                return hugepage_mode::explicit_allocation;
+            std::fprintf(stderr, "warning: env variable GT_HUGEPAGE_MODE set to invalid value '%s'\n", env_value);
+            return hugepage_mode::transparent;
+        }
+
+        inline std::size_t hugepage_size() {
+            std::size_t default_value = 2 * 1024 * 1024; // 2MB is the default on most systems
+#ifdef __linux__
+            auto *fp = std::fopen("/proc/meminfo", "r");
+            if (fp) {
+                char *line = nullptr;
+                size_t line_length;
+                while (getline(&line, &line_length, fp) != -1) {
+                    if (sscanf(line, "Hugepagesize: %lu kB", &default_value) == 1) {
+                        default_value *= 1024;
+                        break;
+                    }
+                }
+                free(line);
+                std::fclose(fp);
+            }
+#endif
+            return default_value;
+        }
+
+        struct ptr_metadata {
+            std::size_t offset, full_size;
+            hugepage_mode mode;
+        };
+
+    } // namespace hugepage_alloc_impl_
 
     /**
      * @brief Allocates huge page memory (if GT_NO_HUGETLB is not defined) and shifts allocations by some bytes to
      * reduce cache set conflicts.
      */
     inline void *hugepage_alloc(std::size_t size) {
-        static std::atomic<std::size_t> s_counter(0);
-        auto offset = 64 << (s_counter++ %7);
+        static std::atomic<std::size_t> s_offset(0);
+        static const std::size_t cache_line_size = hugepage_alloc_impl_::cache_line_size();
+        static const std::size_t cache_sets = hugepage_alloc_impl_::cache_sets();
+        static const std::size_t hugepage_size = hugepage_alloc_impl_::hugepage_size();
+#ifdef __linux__
+        static const std::size_t page_size = sysconf(_SC_PAGESIZE);
+#else
+        static const std::size_t page_size = 4096; // just assume 4KB, default on most systems
+#endif
+        assert(cache_line_size >= sizeof(hugepage_alloc_impl_::ptr_metadata));
+
+        std::size_t offset =
+            ((s_offset++ % cache_sets) << hugepage_alloc_impl_::ilog2(cache_line_size)) + cache_line_size;
+
+        std::size_t full_size = size + offset;
 
-        void *ptr;
-        if (posix_memalign(&ptr, 2 * 1024 * 1024, size + offset))
-            throw std::bad_alloc();
+        void *ptr = nullptr;
+        auto mode = hugepage_alloc_impl_::hugepage_mode_from_env();
+        switch (mode) {
+        case hugepage_alloc_impl_::hugepage_mode::disabled:
```

Comment:
The `mode` could be part of the `hugepage_alloc_impl_`, and the case statement too, in a function that returns a pointer or throws `std::bad_alloc()`.

Reply:
Done.